### DISCUSSION TITLE
transform hyphenated string to camelCase

### DISCUSSION THREAD

**Discussion by jonathanstanley** [colinhacks/zod#2240] [Q&A] [answered]
Wonderful library thank you!

Others may come across this:
* [HTTP headers are case insensitive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers). Therefore headers typically use hyphenation (ex: `my-property-here`).
* [Javascript requires brackets for hyphenated properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Working_with_Objects#accessing_properties), therefore properties typically use camelCase (ex: `myPropertyHere`).

It would be nice to have a simple function that transforms between the two case systems. Even better if it retains description comments.

ex:
```javascript
z.object({
  /** the user's  global id */
  "user-id": z.string().startsWith("gid"),
   /** the type of the request body */
  "content-type": z.string(),
}).transformKeys('toCamelCase');

// renders type {userId:string, contentType:string}
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
import { mapKeys, camel } from 'radash'

const schema = z.record( z.string() )
    .transform( x => mapKeys( x, camel ) )
    .pipe(
        z.object( {
            userId: z.string().startsWith( 'gid' ),
            contentType: z.string(),
        } )
    )
type Data = z.infer<typeof schema>
// type Data = {
//     userId: string
//     contentType: string
// }

console.log( schema.parse( {
    'user-id': 'gid123',
    'content-type': 'text'
} ) )
// { userId: 'gid123', contentType: 'text' }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by jonathanstanley** [reply]
very nice suggestion! i do like the [radash string formatters](https://github.com/rayepps/radash/blob/31c1397437d7fb7a78e97499c8d46f992c49844c/src/string.ts#L20-L31)

**Comment by mengxi-ream** [comment]
You can extract it as a function:

```ts
import { camel, mapKeys } from "radash";
import { z } from "zod";

export const camelCaseSchemaDef = <T extends z.ZodTypeAny>(schema: T) =>
  z
    .record(z.any())
    .transform((x) => mapKeys(x, camel))
    .pipe(schema) as T;

```

Use it like:

```ts
export const summarySchema = camelCaseSchemaDef(
  z.object({
    isArticle: z.boolean(),
    summary: z.string(),
    introduction: z.string(),
    terms: z.array(z.string()),
  })
);

type Summary = z.infer<typeof summarySchema>
// type Summary = {
//     isArticle: z.boolean(),
//     summary: string,
//.    introduction: z.string(),
//.    terms: z.array(z.string()),
// }

summarySchema.parse({
    is_article: true,
    summary: "abc",
    introduction: "abc",
    terms: ["abc", "bca"],
})
```
